crates/misc/component-fuzz-util/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+
+const MAX_FLAT_PARAMS: usize = 16;
+const MAX_FLAT_RESULTS: usize = 1;
+const MAX_ARITY: usize = 5;
+
+/// The name of the imported host function which the generated component will call
+pub const IMPORT_FUNCTION: &str = "echo";
+
+/// The name of the exported guest function which the host should call
+pub const EXPORT_FUNCTION: &str = "echo";
+
+/// Maximum length of an arbitrary tuple type.  As of this writing, the `wasmtime::component::func::typed` module
+/// only implements the `ComponentType` trait for tuples up to this length.
+const MAX_TUPLE_LENGTH: usize = 16;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum CoreType {
+    I32,
+    I64,
+    F32,
+    F64,
+}
+
+impl CoreType {
+    /// This is the `join` operation specified in [the canonical
+    /// ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening) for
+    /// variant types.
+    fn join(self, other: Self) -> Self {
+        match (self, other) {
+            _ if self == other => self,
+            (Self::I32, Self::F32) | (Self::F32, Self::I32) => Self::I32,
+            _ => Self::I64,
+        }
+    }
+}
+
+impl fmt::Display for CoreType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::I32 => f.write_str("i32"),
+            Self::I64 => f.write_str("i64"),
+            Self::F32 => f.write_str("f32"),
+            Self::F64 => f.write_str("f64"),
+        }
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices
+#[derive(Debug)]
+pub struct NonEmptyArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for NonEmptyArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
+                .chain(input.arbitrary_iter()?)
+                .collect::<arbitrary::Result<_>>()?,
+        ))
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices of length
+/// less than or equal to the longest tuple for which Wasmtime generates a `ComponentType` impl
+#[derive(Debug)]
+pub struct TupleArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for TupleArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
+                .chain(input.arbitrary_iter()?)
+                .take(MAX_TUPLE_LENGTH)
+                .collect::<arbitrary::Result<_>>()?,
+        ))
+    }
+}
+
+/// Represents a component model interface type
+#[allow(missing_docs)]
+#[derive(Arbitrary, Debug)]
+pub enum Type {
+    Unit,
+    Bool,
+    S8,
+    U8,
+    S16,
+    U16,
+    S32,
+    U32,
+    S64,
+    U64,
+    Float32,
+    Float64,
+    Char,
+    String,
+    List(Box<Type>),
+    Record(NonEmptyArray<Type>),
```

Comment:
I think we'll want to allow this to have zero fields as well as `Flags` below to exercise the zero-size case.